javascripts/2048.js

```diff
@@ -1,32 +1,406 @@
+    if(arrow_keys.indexOf(direction) > -1) {
+      makeTurn(direction);
       event.preventDefault();
     }
   })
+
+  $('.button.new-game').click(function() {
+    location.reload();
+  })
 })
 
-function moveTile(tile, direction) {
-  var new_tile_value = tile.attr("data-val") * 2;
-  tile.attr("data-val", new_tile_value);
-  tile.text(new_tile_value);
-
-  switch(direction) {
-    case 38: //up
-      tile.attr("data-row","r0");
-      break;
-    case 40: //down
-      tile.attr("data-row","r3");
-      break;
-    case 37: //left
-      tile.attr("data-col","c0");
-      break;
-    case 39: //right
-      tile.attr("data-col","c3");
-      break;
+function prepareBoard() {
+  addTile();
+  addTile();
+}
+
+
+function tileSelectorText(row, col) {
+  return ".tile[data-row=\"" + row + "\"][data-col=\"" + col + "\"]";
+}
+
+function collectEmptySpaces() {
+  var emptySpaces = [];
+  var row = null;
+  var col = null;
+  var tileLocation = null;
+  // figure out all empty spots
+  for (var i = 1; i < 5; i++) { // loop through rows
+    row = i;
+    for (var j = 1; j < 5; j++) { // loop through cols
+      col = j;
+      tileLocation = $(tileSelectorText(row, col));
+      if (tileLocation.length == 0) {
+        emptySpaces.push([row, col]);
+      }
+    }
+  }
+  return emptySpaces;
+}
+
+function addTile() {
+  var tile = $("<div data-row='' data-col='' data-val=''></div>");
+  var dataVal = Math.random() < 0.04 ? 4 : 2;
+  var emptySpaces = collectEmptySpaces(); // guard for if no empty spaces?
+
+  // function to pick row and column number
+  var randomLocation = emptySpaces[Math.floor(Math.random() * emptySpaces.length)];
+
+  tile.addClass("tile");
+  tile.attr("data-row", randomLocation[0]);
+  tile.attr("data-col", randomLocation[1]);
+  tile.attr("data-val", dataVal);
+  tile.text(dataVal);
+  tile.addClass("new");
+
+  $("#gameboard").append(tile);
+  // remove new class after animation is complete
+  tile.on("animationend", function() {
+    tile.removeClass("new");
+  })
+}
+
+function makeTurn(direction) {
+  var tiles = $(".tile"); // in order by the html
+  var turnType = rowOrColumn(direction);
+  var turnOppositeType = (turnType == "data-row") ? "data-col" : "data-row";
+  // set the positive or negative magnitude according to direction
+  var turnMagnitude = parseInt(getMagnitude(direction));
+  var score = parseInt($("#score").attr("data-score"));
+
+  // take snapshot of board at start of turn
+  var startingGameboard = gameboardSnapshot($(".tile"));
+
+  function gameboardSnapshot(tiles) {
+    var gameboard = [];
+    for (var i = 0; i < tiles.length; i++) {
+      var dataRow = tiles[i].getAttribute("data-row");
+      var dataCol = tiles[i].getAttribute("data-col");
+      var dataVal = tiles[i].getAttribute("data-val");
+      var tileProperties = new Array(dataRow, dataCol, dataVal);
+      gameboard[i] = tileProperties;
+    }
+    return gameboard;
+  }
+
+  function validMove() {
```

Comment:
This is an interesting approach. Comparing the initial snapshot with new the state of the game to know whether a move happened is a bit expensive. In this case with 4 rows and 4 columns it is not an issue at all, but a nice optimization would be to keep track on each turn whether it triggered a move or not.